packages/core/src/input/InputManager.ts

```diff
@@ -0,0 +1,320 @@
+            this._activePointerCount === 1 && (evtTypeList = (evtTypeList << 1) | PointerEventType.Up);
+            break;
+          case "pointermove":
+            this._changePointer(PointerOperationType.Update, evt.pointerId, evt.pageX, evt.pageY);
+            break;
+          case "pointercancel":
+          case "pointerout":
+            this._changePointer(PointerOperationType.Remove, evt.pointerId);
+            break;
+          default:
+            break;
+        }
+      }
+      /** Reset event list. */
+      _eventList.length = 0;
+    }
+
+    /** Sync _pointers and _pointerList. */
+    const { _pointers: pointers, _pointerList: pointerList, _input: input, _activePointerCount } = this;
+    pointers.length = _activePointerCount;
+    if (evtTypeList <= 1 && _activePointerCount === 0) {
+      return;
+    } else {
+      for (let i = 0; i < _activePointerCount; i++) {
+        pointers[i] = pointerList[i];
+      }
+    }
+
+    /** Check whether pressed events are triggered. */
+    let prePressedEntity = input.pressedEntity;
+    if (prePressedEntity) {
+      const preScripts = prePressedEntity._scripts._elements;
+      for (let i = preScripts.length - 1; i >= 0; i--) {
+        preScripts[i].onPointerDrag();
+      }
+    }
+    /** Get the entity hit by the ray. */
+    const curEntity = this._pointerRaycast(input.x, input.y);
+    const preEnteredEntity = input.enteredEntity;
+    /** 80% of operations are on curEntity, so we cache his scripts. */
+    let curEntityScripts: Script[];
+    let curEntityScriptsLen = 0;
+    if (curEntity) {
+      curEntityScripts = curEntity._scripts._elements;
+      curEntityScriptsLen = curEntityScripts.length;
+    }
+    /** Check whether enter and exit events are triggered. */
+    if (preEnteredEntity != curEntity) {
+      if (curEntity) {
+        for (let i = 0; i < curEntityScriptsLen; i++) {
+          curEntityScripts[i].onPointerEnter();
+        }
+      }
+      if (preEnteredEntity) {
+        const scripts = preEnteredEntity._scripts;
+        const scriptsArr = scripts._elements;
+        for (let i = scripts.length - 1; i >= 0; i--) {
+          scriptsArr[i].onPointerExit();
+        }
+      }
+      input.enteredEntity = curEntity;
+    }
+    /** Check whether down up and click events are triggered. */
+    while (evtTypeList > 1) {
+      switch (evtTypeList % 2) {
+        case PointerEventType.Down:
+          if (curEntity) {
+            for (let j = 0; j < curEntityScriptsLen; j++) {
+              curEntityScripts[j].onPointerDown();
+            }
+          }
+          prePressedEntity = curEntity;
+          break;
+        case PointerEventType.Up:
+          if (prePressedEntity) {
+            if (prePressedEntity == curEntity) {
+              for (let j = 0; j < curEntityScriptsLen; j++) {
+                const script = curEntityScripts[j];
+                script.onPointerUp();
+                script.onPointerClick();
```

Comment:
`onPointerUp `  after `onPointerClick ` is better,because Down is start,Up is end.

Reply:
Done.